igvm/vm.py

```diff
@@ -335,6 +475,106 @@ def build(self, run_puppet=True, debug_puppet=False, postboot=None):
+                SubnetId=self.dataset_obj['aws_subnet_id'],
+                Placement={
+                    'AvailabilityZone': self.dataset_obj['aws_placement']
+                },
+                PrivateIpAddress=str(self.dataset_obj['intern_ip']),
+                UserData='' if postboot is None else postboot,
+                TagSpecifications=[
+                    {
+                        'ResourceType': 'instance',
+                        'Tags': [
+                            {
+                                'Key': 'Name',
+                                'Value': self.dataset_obj['hostname'],
+                            },
+                        ]
+                    },
+                ],
+                DryRun=False,
+                MinCount=1,
+                MaxCount=1)
+            log.info(response)
+        except ClientError as e:
+            raise VMError(e)
+
+        if run_puppet:
+            self.run_puppet(clear_cert=True, debug=debug_puppet)
+
+        self.dataset_obj['aws_instance_id'] = response['Instances'][0][
+            'InstanceId']
+
+        log.info('waiting for {} to be started'.format(
+            self.dataset_obj['hostname']))
+        vm_setup = tqdm.tqdm(
+            total=timeout_vm_setup, desc='vm_setup', position=0)
+        cloud_init = tqdm.tqdm(
+            total=timeout_cloud_init, desc='cloud_init', position=1)
+        vm_status_code = self.aws_describe_instance_status(
+            self.dataset_obj['aws_instance_id'])
+        for retry in range(timeout_vm_setup):
+            vm_setup.update(1)
+            vm_status_code = self.aws_describe_instance_status(
+                self.dataset_obj['aws_instance_id'])
+            if vm_status_code == AWS_RETURN_CODES['running']:
+                vm_setup.update(timeout_vm_setup - retry - 1)
+                for retry in range(timeout_cloud_init):
+                    cloud_init.update(1)
+                    with settings(host_string=self.dataset_obj['hostname'],
+                                  warn_only=True, abort_on_prompts=True):
+                        # Only try to connect every 10s to avoid
+                        # paramiko exceptions
+                        if retry % 10 == 0:
+                            try:
+                                with hide('aborts'):
+                                    command = run((
+                                        'find /var/lib/cloud/instance/'
+                                        'boot-finished'), quiet=True)
+                                if command.succeeded:
+                                    cloud_init.update(
+                                        timeout_cloud_init - retry - 1)
+                                    break
+                            except (SystemExit, NetworkError) as e:
+                                time.sleep(1)
+                        else:
+                            time.sleep(1)
+                break
+            time.sleep(1)
+
```

Comment:
May I offer the slightly more readable:

```python3

        # Wait for AWS to decalre the VM running
        while (
            timeout_vm_setup and
            AWS_RETURN_CODES['running'] != self.aws_describe_instance_status(
                self.dataset_obj['aws_instance_id']
            )
        ):
            vm_setup.update(1)
            timeout_vm_setup -= 1
            time.sleep(1)
        vm_setup.update(timeout_vm_setup - 1)
        # TODO: Handle overrun timeout

        # Try to provision the VM
        for retry in range(timeout_cloud_init):
            cloud_init.update(1)

            # Only try to connect every 10s to avoid paramiko exceptions
            if retry % 10 != 0:
                time.sleep(1)
                continue

            with settings(
                hide('aborts'),
                host_string=self.dataset_obj['hostname'],
                warn_only=True,
                abort_on_prompts=True,
            ):
                try:
                    if run(
                        'find /var/lib/cloud/instance/ boot-finished',
                        quiet=True
                    ).succeeded:
                        cloud_init.update(timeout_cloud_init - retry - 1)
                        break
                except (SystemExit, NetworkError):
                    time.sleep(1)
        # TODO: Handle overrun timeout
```

Reply:
Saving you four levels of indent.

Reply:
Looks promising, will check it! Thank you

Reply:
Worked like a charme, added!